From: Arturo Buzarra <arturo.buzarra@digi.com>
Date: Mon, 21 Jan 2019 11:30:57 +0100
Subject: [PATCH] networkmanager: trigger dispatcher on per-device connectivity
 loss

This creates a new dispatcher action DEVICE_CONNECTIVITY_CHANGE, that
gets triggered whenever one interface fails in the upstream connectivity
check, regardless of the system having connectivity through a different
interface.

Signed-off-by: Javier Viguera <javier.viguera@digi.com>
Signed-off-by: Arturo Buzarra <arturo.buzarra@digi.com>
---
 shared/nm-libnm-core-aux/nm-dispatcher-api.h |  1 +
 src/devices/nm-device.c                      |  4 ++
 src/nm-dispatcher.c                          | 40 +++++++++++++++++++-
 src/nm-dispatcher.h                          | 10 ++++-
 4 files changed, 53 insertions(+), 2 deletions(-)

diff --git a/shared/nm-libnm-core-aux/nm-dispatcher-api.h b/shared/nm-libnm-core-aux/nm-dispatcher-api.h
index e6d0d92..123d737 100644
--- a/shared/nm-libnm-core-aux/nm-dispatcher-api.h
+++ b/shared/nm-libnm-core-aux/nm-dispatcher-api.h
@@ -53,6 +53,7 @@
 #define NMD_ACTION_DHCP4_CHANGE "dhcp4-change"
 #define NMD_ACTION_DHCP6_CHANGE "dhcp6-change"
 #define NMD_ACTION_CONNECTIVITY_CHANGE "connectivity-change"
+#define NMD_ACTION_DEVICE_CONNECTIVITY_CHANGE "device-connectivity-change"

 typedef enum {
 	DISPATCH_RESULT_UNKNOWN = 0,
diff --git a/src/devices/nm-device.c b/src/devices/nm-device.c
index 43f0373..b99fd41 100644
--- a/src/devices/nm-device.c
+++ b/src/devices/nm-device.c
@@ -2853,6 +2853,10 @@ concheck_update_state (NMDevice *self,
 	_LOGD (LOGD_CONCHECK, "connectivity state changed from %s to %s",
 	       nm_connectivity_state_to_string (priv->concheck_x[IS_IPv4].state),
 	       nm_connectivity_state_to_string (state));
+
+	if (priv->concheck_x[IS_IPv4].state == NM_CONNECTIVITY_FULL)
+		nm_dispatcher_call_device_connectivity(state, self, NULL, NULL, NULL);
+
 	priv->concheck_x[IS_IPv4].state = state;

 	_notify (self, IS_IPv4 ? PROP_IP4_CONNECTIVITY : PROP_IP6_CONNECTIVITY);
diff --git a/src/nm-dispatcher.c b/src/nm-dispatcher.c
index d656edc..39b781c 100644
--- a/src/nm-dispatcher.c
+++ b/src/nm-dispatcher.c
@@ -464,7 +464,8 @@ static const char *action_table[] = {
 	[NM_DISPATCHER_ACTION_VPN_DOWN]     = NMD_ACTION_VPN_DOWN,
 	[NM_DISPATCHER_ACTION_DHCP4_CHANGE] = NMD_ACTION_DHCP4_CHANGE,
 	[NM_DISPATCHER_ACTION_DHCP6_CHANGE] = NMD_ACTION_DHCP6_CHANGE,
-	[NM_DISPATCHER_ACTION_CONNECTIVITY_CHANGE] = NMD_ACTION_CONNECTIVITY_CHANGE
+	[NM_DISPATCHER_ACTION_CONNECTIVITY_CHANGE] = NMD_ACTION_CONNECTIVITY_CHANGE,
+	[NM_DISPATCHER_ACTION_DEVICE_CONNECTIVITY_CHANGE] = NMD_ACTION_DEVICE_CONNECTIVITY_CHANGE
 };

 static const char *
@@ -912,6 +913,43 @@ nm_dispatcher_call_connectivity (NMConnectivityState connectivity_state,
 	                         callback, user_data, out_call_id);
 }

+/**
+ * nm_dispatcher_call_device_connectivity():
+ * @connectivity_state: the #NMConnectivityState value
+ * @device: the #NMDevice the action applies to
+ * @callback: a caller-supplied callback to execute when done
+ * @user_data: caller-supplied pointer passed to @callback
+ * @out_call_id: on success, a call identifier which can be passed to
+ * nm_dispatcher_call_cancel()
+ *
+ * This method does not block the caller.
+ *
+ * Returns: %TRUE if the action was dispatched, %FALSE on failure
+ */
+gboolean
+nm_dispatcher_call_device_connectivity(NMConnectivityState connectivity_state,
+				       NMDevice *device,
+				       NMDispatcherFunc callback,
+				       gpointer user_data, guint *out_call_id)
+{
+	NMActRequest *act_request;
+
+	nm_assert(NM_IS_DEVICE(device));
+
+	act_request = nm_device_get_act_request(device);
+	if (!act_request)
+		return FALSE;
+
+	nm_assert(NM_IN_SET(nm_active_connection_get_device(NM_ACTIVE_CONNECTION(act_request)), NULL, device));
+	return _dispatcher_call(NM_DISPATCHER_ACTION_DEVICE_CONNECTIVITY_CHANGE, FALSE, device,
+				nm_act_request_get_settings_connection(act_request),
+				nm_act_request_get_applied_connection(act_request),
+				nm_active_connection_get_activation_type(NM_ACTIVE_CONNECTION(act_request)) == NM_ACTIVATION_TYPE_EXTERNAL,
+				connectivity_state,
+				NULL, NULL, NULL, NULL,
+				callback, user_data, out_call_id);
+}
+
 void
 nm_dispatcher_call_cancel (guint call_id)
 {
diff --git a/src/nm-dispatcher.h b/src/nm-dispatcher.h
index 1cdeeb8..9902a77 100644
--- a/src/nm-dispatcher.h
+++ b/src/nm-dispatcher.h
@@ -36,7 +36,8 @@ typedef enum {
 	NM_DISPATCHER_ACTION_VPN_DOWN,
 	NM_DISPATCHER_ACTION_DHCP4_CHANGE,
 	NM_DISPATCHER_ACTION_DHCP6_CHANGE,
-	NM_DISPATCHER_ACTION_CONNECTIVITY_CHANGE
+	NM_DISPATCHER_ACTION_CONNECTIVITY_CHANGE,
+	NM_DISPATCHER_ACTION_DEVICE_CONNECTIVITY_CHANGE
 } NMDispatcherAction;

 typedef void (*NMDispatcherFunc) (guint call_id, gpointer user_data);
@@ -82,6 +83,13 @@ gboolean nm_dispatcher_call_connectivity (NMConnectivityState state,
                                           gpointer user_data,
                                           guint *out_call_id);

+gboolean nm_dispatcher_call_device_connectivity(NMConnectivityState
+                                                connectivity_state,
+                                                NMDevice *device,
+                                                NMDispatcherFunc callback,
+                                                gpointer user_data,
+                                                guint *out_call_id);
+
 void nm_dispatcher_call_cancel (guint call_id);

 void nm_dispatcher_init (void);
